Provide additional clarifications around OCB (#6714)
open-telemetry/opentelemetry.io: content/en/docs/collector/custom-collector.md

## Step 1 - Install the builder

{{% alert color="primary" title="NOTE" %}}

> The `ocb` tool requires Go to build the Collector distribution. Please
> [install Go](https://go.dev/doc/install) on your machine, if you haven't done so

Reviewer: ```suggestion
The `ocb` tool requires Go to build the Collector distribution. [Install Go](https://go.dev/doc/install) on your machine, if you haven't done so
```

{{% /alert %}}

> ## Step 3a - Generate the Code and Build your Collector's distribution

Reviewer: ```suggestion
## Step 3a - Generate the code and build your Collector's distribution
```

{{% alert color="primary" title="NOTE" %}}

This step is used to build your custom collector distribution using the `ocb`

> binary. If you would like to build and deploy your custom collector distribution
> to a container orchestrator (e.g. Kubernetes), skip this step and go to

Reviewer: ```suggestion
binary. If you would like to build and deploy your custom Collector distribution
to a container orchestrator (for example, Kubernetes), skip this step and go to
```

projects and easily build and distribute your own collector distribution with
your components.

> ## Step 3b - Containerizing your Collector's distribution

Reviewer: ```suggestion
## Step 3b - Containerize your Collector's distribution
```

{{% alert color="primary" title="NOTE" %}}

> This step will build your collector distribution in-place inside a `Dockerfile`.

Reviewer 2: What does in place mean in this context?

Author: @jaronoff97 I mean that it's doing it inside the Dockerfile, rather than outside of it. But if that's confusing, I can totally remove that wording.

> Follow this step if you need to deploy your collector distribution to a
> container orchestrator (e.g. Kubernetes). If you would like to _only_ build your

Reviewer: ```suggestion
Follow this step if you need to deploy your Collector distribution to a
container orchestrator (for example, Kubernetes). If you would like to _only_ build your
```

collector distribution without containerization, go to
[Step 3a](#step-3a---generate-the-code-and-build-your-collectors-distribution).

{{% /alert %}}

> You will need to add two new files to your project:
> 
> - `Dockerfile` - Container image definition of your collector distribution
> - `collector-config.yaml` - Minimalist collector configuration YAML for testing
>   your distribution
> 
> Your file structure will look like this:

Reviewer: ```suggestion
You need to add two new files to your project:

- `Dockerfile` - Container image definition of your Collector distribution
- `collector-config.yaml` - Minimalist Collector configuration YAML for testing
  your distribution

After adding these files, your file structure looks like this:
```

```console
.
├── builder-config.yaml
├── collector-config.yaml
└── Dockerfile
```

> The `Dockerfile` below will build you collector distribution in-place, ensuring
> that the resulting collector distribution binary matches the target container
> architecture (e.g. `linux/arm64`, `linux/amd64`):

Reviewer: ```suggestion
The following `Dockerfile` builds your Collector distribution in-place, ensuring
that the resulting Collector distribution binary matches the target container
architecture (for example, `linux/arm64`, `linux/amd64`):
```

<!-- prettier-ignore-start -->
```yaml
FROM alpine:3.19 AS certs
RUN apk --update add ca-certificates

FROM golang:1.23.6 AS build-stage
WORKDIR /usr/bin/otelcol

COPY ./builder-config.yaml builder-config.yaml
COPY ./otelcol-config.yaml config.yaml

RUN --mount=type=cache,target=/root/.cache/go-build GO111MODULE=on go install go.opentelemetry.io/collector/cmd/builder@{{% version-from-registry collector-builder %}}
RUN mkdir -p ./otelcol-dev && chmod +x ./otelcol-dev
RUN --mount=type=cache,target=/root/.cache/go-build builder --config builder-config.yaml

FROM gcr.io/distroless/base:latest

ARG USER_UID=10001
USER ${USER_UID}

COPY --from=build-stage /usr/bin/otelcol/config.yaml /etc/otelcol-contrib/config.yaml
COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
COPY --chmod=755 --from=build-stage /usr/bin/otelcol/otelcol-dev /otelcol

ENTRYPOINT ["/otelcol/otelcol-dev"]
CMD ["--config", "/etc/otelcol-contrib/config.yaml"]

> EXPOSE 4317 4318 12001 55680 55679

Reviewer 3: I think it would be good to have people more experienced than me with Docker review this Dockerfile before we start recommending it in our docs.

A few things I noticed:

- The second COPY statement should probably reference `collector-config.yaml` instead of `otelcol-config.yaml` to match the rest of the tutorial. Moreover, since I don't think OCB needs this file at all, we should probably COPY it directly into the final image instead of passing it through `build-stage`.

- Instead of using a `golang` image and `go install`ing OCB, would there be an issue with directly using the OCB Docker images we release (`ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector-builder`)? I'll admit I haven't tried it.

- (Nitpick) Are the 12001 and 55680 ports used for anything? And 55679 could probably be removed since the zpagesextension isn't included in this example distro.

- (Nitpick) Maybe update the reference to `otelcol-contrib` in the config file's path to something else?

Author: Maybe @jaronoff97 can address some of the questions, since I got the Dockerfile from him. I did make a few of the updates that you suggested, because you're right that some of these don't make sense. I had added the additional ports based on [this Dockerfile in the opentelemetry-collector-contrib repo](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c2e12729a13e4cef07b503fede311cb8c8942af9/cmd/otelcontribcol/Dockerfile#L12). I didn't use that particular Dockerfile because it doesn't run the OCB inside the Dockerfile, which I personally think is a better idea since that way you can do multi-arch builds more easily and elegantly.

```
<!-- prettier-ignore-end -->

> Below is the minimalist `collector-config.yaml` definition:

Reviewer: ```suggestion
The following is the minimalist `collector-config.yaml` definition:
```

```yaml
receivers:
  otlp:
    protocols:
      grpc: {}
      http: {}

processors:
  batch: {}

exporters:
  debug:
    verbosity: detailed

service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [debug]
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [debug]
    logs:
      receivers: [otlp]
      processors: [batch]
      exporters: [debug]
```

> Use the commands below to build a multi-architecture Docker image of the OCB
> using `linux/amd64` and `linux/arm64` as the target build architectures. Learn
> more about multi-architecture builds
> [here](https://blog.jaimyn.dev/how-to-build-multi-architecture-docker-images-on-an-m1-mac/).

Reviewer: ```suggestion
Use the following commands to build a multi-architecture Docker image of the OCB
using `linux/amd64` and `linux/arm64` as the target build architectures. To learn more, 
see this [blog post](https://blog.jaimyn.dev/how-to-build-multi-architecture-docker-images-on-an-m1-mac/) about multi-architecture builds.
```

```console
.
├── builder-config.yaml
├── collector-config.yaml
└── Dockerfile
```

The following `Dockerfile` builds your Collector distribution in-place, ensuring
that the resulting Collector distribution binary matches the target container
architecture (for example, `linux/arm64`, `linux/amd64`):

<!-- prettier-ignore-start -->
```yaml
FROM alpine:3.19 AS certs
RUN apk --update add ca-certificates

FROM golang:1.23.6 AS build-stage
WORKDIR /usr/bin/otelcol

COPY ./builder-config.yaml builder-config.yaml

RUN --mount=type=cache,target=/root/.cache/go-build GO111MODULE=on go install go.opentelemetry.io/collector/cmd/builder@{{% version-from-registry collector-builder %}}
RUN mkdir -p ./otelcol-dev && chmod +x ./otelcol-dev
RUN --mount=type=cache,target=/root/.cache/go-build builder --config builder-config.yaml

FROM gcr.io/distroless/base:latest

ARG USER_UID=10001
USER ${USER_UID}

COPY ./otelcol-config.yaml /etc/otelcol-contrib/otelcol-config.yaml
COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
COPY --chmod=755 --from=build-stage /usr/bin/otelcol/otelcol-dev /otelcol

ENTRYPOINT ["/otelcol/otelcol-dev"]

> CMD ["--config", "/etc/otelcol-contrib/otelcol-config.yaml"]

Author: I tested with your changes, and it seems to work!

EXPOSE 4317 4318 12001
```
<!-- prettier-ignore-end -->

The following is the minimalist `collector-config.yaml` definition:

```yaml
receivers:
  otlp:
    protocols:

>       grpc: {}
>       http: {}
> 
> processors:
>   batch: {}

Reviewer 3: The default endpoints for the OTLP receiver are `localhost:4317/4318`, which I don't think is going to work in Docker unless you put the application emitting the telemetry inside the same container. Also, referring to [the configuration suggested on this page](https://opentelemetry.io/docs/collector/configuration/#basics), the `{}` doesn't seem necessary.
```suggestion
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
processors:
  batch:
```

Author: Personally didn't have any issues when I tested with my previous file, but I'm fine with these changes.

Reviewer 3: Oh, that's surprising. In my tests, when running "docker run -p 4317:4317 ocb-test", it was impossible to connect to port 4317 unless I made that change to the config.

Maybe you were using an older Collector version? We [changed the default](https://github.com/open-telemetry/opentelemetry-collector/pull/10352) from 0.0.0.0 to localhost in v0.104.0.
